README.txt
# bash-tools
bashtest/ - bash testing library; assert, diff, stack trace, etc

genbash/  - generate bash boilerplate with argument parsing and error reporting
code in place

prep/     - generate strings with positional arguments from the command line,
e.g. 
$ echo host port | bash prep.sh -s 'nc -vz #1 #2'
nc -vz host port


prep/changelog.txt
2021-01-15
cmd-gen.sh 0.9 -> prep.sh 0.91
syntax check regex is now checked for.. well, syntax
field separator option added
positional argument 0 is now the whole line

2021-01-20
prep.sh 0.91 -> prep.sh 1.0
test finished

2021-01-21
readme.txt and fd input test added

2021-03-08
prep.sh 1.0 -> prep.sh 1.1
default 0 to 9 positional arguments limitation eliminated by expanding backwards
e.g. now '#10' expands before '#1', so the result is <arg_10>, and not <arg_1>0


prep/test/run-tests.sh
#!/bin/bash

G_BASE_DIR="$(dirname $(realpath $0))"

function run_prep
{
	bash ../prep.sh "$@"
}

function main
{
	source "$G_BASE_DIR/bashtest.sh"
	
	if [ "$#" -gt 0 ]; then
		bt_set_verbose
	fi
	
	bt_enter
	test_err
	test_ok
	bt_exit_success
	
}

function test_err
{

local L_GOT=""
L_GOT="$(run_prep 2>&1 || \
run_prep '-f' 2>&1 || run_prep '-F' 2>&1 || run_prep '-p' 2>&1 || run_prep '-t' 2>&1 || \
run_prep '-s' 2>&1 || run_prep '-c' 2>&1 || run_prep '-d' 2>&1 || \
run_prep '--fields' 2>&1 || run_prep '--field-sep' 2>&1 || run_prep '--pos-spec' 2>&1 || \
run_prep '--syntax-str' 2>&1 || run_prep '--string' 2>&1 || run_prep '--syntax-check' 2>&1 || \
run_prep '--dry-run' 2>&1 || run_prep -u 2>&1 || run_prep --unknown 2>&1)"
	bt_assert_failure
	
local L_EXP="Use: prep.sh <option> [args..] [files]
Try 'prep.sh --help' for help
../prep.sh: error: '-f' missing argument
../prep.sh: error: '-F' missing argument
../prep.sh: error: '-p' missing argument
../prep.sh: error: '-t' missing argument
../prep.sh: error: '-s' missing argument
../prep.sh: error: '-c' missing argument
../prep.sh: error: '--string' not given
../prep.sh: error: '--fields' missing argument
../prep.sh: error: '--field-sep' missing argument
../prep.sh: error: '--pos-spec' missing argument
../prep.sh: error: '--syntax-str' missing argument
../prep.sh: error: '--string' missing argument
../prep.sh: error: '--syntax-check' missing argument
../prep.sh: error: '--string' not given
../prep.sh: error: '-u' unknown option
../prep.sh: error: '--unknown' unknown option"
	bt_diff "<(echo \"$L_GOT\")" "<(echo \"$L_EXP\")"
	bt_assert_success
	
L_GOT="$(run_prep -s str -c 'foo' 2>&1)"
	bt_assert_failure
L_EXP='../prep.sh: error: "foo" should match ^[0-9]+~.+$, but does not'
	bt_diff "<(echo \"$L_GOT\")" "<(echo \"$L_EXP\")"
	bt_assert_success
}

function test_ok
{
	local L_GOT=""
	L_GOT="$(echo host port | run_prep -s 'nc -vz #1 #2')"
	bt_assert_success
	
	# trivial
	local L_EXP="nc -vz host port"
	bt_diff "<(echo \"$L_GOT\")" "<(echo \"$L_EXP\")"
	bt_assert_success

	# field sep + multiple lines
	local L_MLN=\
"$(echo 'foo;bar;baz'; echo '#'; echo 'zig;zag;zog'; echo ""; echo '1;2;3'; echo "")"
	L_GOT="$(echo "$L_MLN" | run_prep -f 3 -F ';' -s 'ls #1 #2 #3 #0 #3 #2 #1')"
	bt_assert_success
	
	L_EXP="ls foo bar baz foo;bar;baz baz bar foo
ls zig zag zog zig;zag;zog zog zag zig
ls 1 2 3 1;2;3 3 2 1"
	bt_diff "<(echo \"$L_GOT\")" "<(echo \"$L_EXP\")"
	bt_assert_success
	
	# help examples short
	L_GOT="$(echo 'a;b' | run_prep -F ';' -s 'ls #0 #1 #2')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a;b a b')"
	bt_assert_success
	
	L_GOT="$(echo a b c d e f g h i j | run_prep -f 10 -s 'ls #1 #10')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a a0')"
	bt_assert_success
	
	L_GOT="$(echo a b c d e f g h i j | run_prep -f 10 -p '#%d#' -s 'ls #1# #10#')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a j')"
	bt_assert_success
	
	L_GOT="$(echo a b c | run_prep -s 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b c": 2 fields expected, but got 3 instead'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b c | run_prep -s 'nc #1 #2' -t '<host> <port>' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b c": 2 fields expected, but got 3 instead; syntax should be "<host> <port>"'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b | run_prep -s 'nc #1 #2' -c '1~^localhost$' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": field 1 "a" should match "^localhost$", but does not'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo localhost b | run_prep -s 'nc #1 #2' -c '1~^localhost$' 2>&1)"
	bt_assert_success
	bt_diff "<(echo '$L_GOT')" "<(echo 'nc localhost b')"
	bt_assert_success
	
	L_GOT="$(echo localhost b | run_prep -s 'nc #1 #2' -c '1~^localhost$;2~^[0-9]+$' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "localhost b": field 2 "b" should match "^[0-9]+$", but does not'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo localhost 8000 | run_prep -s 'nc #1 #2' -c '1~^localhost$;2~^[0-9]+$' 2>&1)"
	bt_assert_success
	bt_diff "<(echo '$L_GOT')" "<(echo 'nc localhost 8000')"
	bt_assert_success
	
	# help examples long
	L_GOT="$(echo 'a;b' | run_prep --field-sep ';' --string 'ls #0 #1 #2')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a;b a b')"
	bt_assert_success
	
	L_GOT="$(echo a b c d e f g h i j | run_prep --fields 10 --string 'ls #1 #10')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a a0')"
	bt_assert_success
	
	L_GOT="$(echo a b c d e f g h i j | run_prep --fields 10 --pos-spec '#%d#' --string 'ls #1# #10#')"
	bt_assert_success
	bt_diff "<(echo \"$L_GOT\")" "<(echo 'ls a j')"
	bt_assert_success
	
	L_GOT="$(echo a b c | run_prep --string 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b c": 2 fields expected, but got 3 instead'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b c | run_prep --string 'nc #1 #2' --syntax-str '<host> <port>' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b c": 2 fields expected, but got 3 instead; syntax should be "<host> <port>"'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b | run_prep --string 'nc #1 #2' --syntax-check '1~^localhost$' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": field 1 "a" should match "^localhost$", but does not'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo localhost b | run_prep --string 'nc #1 #2' --syntax-check '1~^localhost$' 2>&1)"
	bt_assert_success
	bt_diff "<(echo '$L_GOT')" "<(echo 'nc localhost b')"
	bt_assert_success
	
	L_GOT="$(echo localhost b | run_prep --string 'nc #1 #2' --syntax-check '1~^localhost$;2~^[0-9]+$' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "localhost b": field 2 "b" should match "^[0-9]+$", but does not'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo localhost 8000 | run_prep --string 'nc #1 #2' --syntax-check '1~^localhost$;2~^[0-9]+$' 2>&1)"
	bt_assert_success
	bt_diff "<(echo '$L_GOT')" "<(echo 'nc localhost 8000')"
	bt_assert_success
	
	# misc
	bt_diff "<(run_prep -v)" "<(echo 'prep.sh 1.0')"
	bt_assert_success
	bt_diff "<(run_prep --version)" "<(echo 'prep.sh 1.0')"
	bt_assert_success
	bt_diff "<(run_prep -s foo -d | wc -l)" "<(echo 109)"
	bt_assert_success
	bt_diff "<(run_prep -s foo --dry-run | wc -l)" "<(echo 109)"
	bt_assert_success
	bt_diff "<(run_prep -h | wc -l)" "<(echo 58)"
	bt_assert_success
	bt_diff "<(run_prep --help | wc -l)" "<(echo 58)"
	bt_assert_success
	
	# non-default field num
	L_GOT="$(echo a b | run_prep -f 3 -s 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": 3 fields expected, but got 2 instead'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b | run_prep --fields 3 -s 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": 3 fields expected, but got 2 instead'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b | run_prep --fields 3 -t '<a> <b> <c>' -s 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": 3 fields expected, but got 2 instead; syntax should be "<a> <b> <c>"'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
	
	L_GOT="$(echo a b | run_prep --fields 3 --syntax-str '<a> <b> <c>' -s 'nc #1 #2' 2>&1)"
	bt_assert_failure
	L_EXP='../prep.sh: error: file "-", line 1: "a b": 3 fields expected, but got 2 instead; syntax should be "<a> <b> <c>"'
	bt_diff "<(echo '$L_GOT')" "<(echo '$L_EXP')"
	bt_assert_success
}

main "$@"


prep/tests/bashtest.sh
#!/bin/bash

# 1.0

set -u

readonly G_BT_SCRIPT_NAME="$(basename $0)"
readonly G_BT_BASE_DIR="$(dirname $(realpath $0))"
readonly G_BT_DEV_NULL="/dev/null"
G_BT_VERBOSE=0

function bt_enter { pushd "$G_BT_BASE_DIR" > "$G_BT_DEV_NULL"; }
function bt_leave { popd > "$G_BT_DEV_NULL"; exit "$1"; }

function bt_exit_success { bt_leave 0; }
function bt_exit_failure { bt_leave 1; }
function bt_exit_failure_strace { bt_stack_trace; bt_exit_failure; }

function bt_assert { bt_eval "$@" || bt_error_exit_st "'$@' failed"; }
function bt_assert_success { bt_assert "[ $? -eq 0 ]"; }
function bt_assert_failure { bt_assert "[ $? -ne 0 ]"; }

function bt_diff { bt_eval "diff $@"; }
function bt_error_print { echo "$G_BT_SCRIPT_NAME: error: $@" >&2; }
function bt_error_exit { bt_error_print "$@"; bt_exit_failure; }
function bt_error_exit_st { bt_error_print "$@"; bt_exit_failure_strace; }

function bt_set_verbose { G_BT_VERBOSE=1; }
function bt_eval { [ "$G_BT_VERBOSE" -ne 1 ] || echo "$@"; eval "$@"; }
function bt_stack_trace
{
	local L_FRAME=0;
	while caller $L_FRAME; do
		((++L_FRAME))
	done
}


prep/prep.sh
#!/bin/bash

# Author: Vladimir Dinev
# vld.dinev@gmail.com
# 2021-03-08

set -u

readonly G_SCRIPT_NAME="$(basename $0)"
readonly G_SCRPIT_DIR="$(dirname $(realpath $0))"
readonly G_SCRIPT_VER="1.1"
function print_version { echo "$G_SCRIPT_NAME $G_SCRIPT_VER"; }

function print_fd2    { echo "$@" >&2; }
function error_print  { print_fd2 "$0: error: $@"; }
function error_exit   { error_print "$@"; exit_failure; }
function exit_success { exit 0; }
function exit_failure { exit 1; }

readonly G_OPT_NUM_FIELDS_S="-f"
readonly G_OPT_NUM_FIELDS_L="--fields"
readonly G_MATCH_NUM_FIELDS="@($G_OPT_NUM_FIELDS_S|$G_OPT_NUM_FIELDS_L)"
G_NUM_FIELDS=""

readonly G_OPT_FIELD_SEP_S="-F"
readonly G_OPT_FIELD_SEP_L="--field-sep"
readonly G_MATCH_FIELD_SEP="@($G_OPT_FIELD_SEP_S|$G_OPT_FIELD_SEP_L)"
G_FIELD_SEP=""

readonly G_OPT_POS_SPEC_S="-p"
readonly G_OPT_POS_SPEC_L="--pos-spec"
readonly G_MATCH_POS_SPEC="@($G_OPT_POS_SPEC_S|$G_OPT_POS_SPEC_L)"
G_POS_SPEC=""

readonly G_OPT_SYNTAX_STR_S="-t"
readonly G_OPT_SYNTAX_STR_L="--syntax-str"
readonly G_MATCH_SYNTAX_STR="@($G_OPT_SYNTAX_STR_S|$G_OPT_SYNTAX_STR_L)"
G_SYNTAX_STR=""

readonly G_OPT_STRING_S="-s"
readonly G_OPT_STRING_L="--string"
readonly G_MATCH_STRING="@($G_OPT_STRING_S|$G_OPT_STRING_L)"
G_STRING=""

readonly G_OPT_SYNTAX_CHECK_S="-c"
readonly G_OPT_SYNTAX_CHECK_L="--syntax-check"
readonly G_MATCH_SYNTAX_CHECK="@($G_OPT_SYNTAX_CHECK_S|$G_OPT_SYNTAX_CHECK_L)"
G_SYNTAX_CHECK=""

readonly G_OPT_DRY_RUN_S="-d"
readonly G_OPT_DRY_RUN_L="--dry-run"
readonly G_MATCH_DRY_RUN="@($G_OPT_DRY_RUN_S|$G_OPT_DRY_RUN_L)"
G_DRY_RUN=""

readonly G_OPT_HELP_S="-h"
readonly G_OPT_HELP_L="--help"
readonly G_MATCH_HELP="@($G_OPT_HELP_S|$G_OPT_HELP_L)"
G_HELP=""

readonly G_OPT_VERSION_S="-v"
readonly G_OPT_VERSION_L="--version"
readonly G_MATCH_VERSION="@($G_OPT_VERSION_S|$G_OPT_VERSION_L)"
G_VERSION=""

G_CMD_LINE_OTHER=""

function set_num_fields { G_NUM_FIELDS="$2"; }
function set_field_sep { G_FIELD_SEP="-F'$2'"; }
function set_pos_spec { G_POS_SPEC="$2"; }
function set_syntax_str { G_SYNTAX_STR="$2"; }
function set_string { G_STRING="$2"; }
function set_syntax_check { G_SYNTAX_CHECK="$2"; }
function set_dry_run { G_DRY_RUN="yes"; }
function set_help { G_HELP="yes"; print_help; }
function set_version { G_VERSION="yes"; print_version; }

function get_args
{
	shopt -s extglob
	local L_UNBOUND_ARG="-*"

	while [ "$#" -gt 0 ]; do
		local L_OPT_ARG=""
		local L_OPT_NO_ARG=""

		case "$1" in
			$G_MATCH_NUM_FIELDS)
				L_OPT_ARG="set_num_fields"
			;;
			$G_MATCH_FIELD_SEP)
				L_OPT_ARG="set_field_sep"
			;;
			$G_MATCH_POS_SPEC)
				L_OPT_ARG="set_pos_spec"
			;;
			$G_MATCH_SYNTAX_STR)
				L_OPT_ARG="set_syntax_str"
			;;
			$G_MATCH_STRING)
				L_OPT_ARG="set_string"
			;;
			$G_MATCH_SYNTAX_CHECK)
				L_OPT_ARG="set_syntax_check"
			;;
			$G_MATCH_DRY_RUN)
				L_OPT_NO_ARG="set_dry_run"
			;;
			$G_MATCH_HELP)
				L_OPT_NO_ARG="set_help"
			;;
			$G_MATCH_VERSION)
				L_OPT_NO_ARG="set_version"
			;;
			$L_UNBOUND_ARG)
				error_exit "'$1' unknown option"
			;;
			*)
				G_CMD_LINE_OTHER="${G_CMD_LINE_OTHER}'$1' "
			;;
		esac

		if [ ! -z "$L_OPT_ARG" ]; then
			if [ "$#" -lt 2 ] || [ "${2:0:1}" == "-" ]; then
				error_exit "'$1' missing argument"
			fi
			eval "$L_OPT_ARG '$1' '$2'"
			shift 2
		elif [ ! -z "$L_OPT_NO_ARG" ]; then
			eval "$L_OPT_NO_ARG '$1'"
			shift
		else
			shift
		fi
	done
}

function print_version
{
	echo "$G_VER_STR"
	exit_success
}

readonly G_VER_STR="$G_SCRIPT_NAME $G_SCRIPT_VER"
readonly G_USE_STR="Use: $G_SCRIPT_NAME <option> [args..] [files]"
function print_use
{
	print_fd2 "$G_USE_STR"
	print_fd2 "Try '$G_SCRIPT_NAME $G_OPT_HELP_L' for help"
	exit_failure
}

function print_help
{
echo "$G_USE_STR"
echo "Generates strings using positional arguments from the command line. Input"
echo "is stdin if no files are given. Empty lines and lines beginning with a"
echo "'#' are ignored."
echo ""
echo "$G_OPT_NUM_FIELDS_S, $G_OPT_NUM_FIELDS_L <num>"
echo "Expect input to have <num> number of fields per line or quit with an"
echo "error. Default is 2."
echo ""
echo "$G_OPT_FIELD_SEP_S, $G_OPT_FIELD_SEP_L <field-sep>"
echo "Passes <field-sep> to awk. E.g."
echo "$ echo 'a;b' | $G_SCRIPT_NAME $G_OPT_FIELD_SEP_S ';' -s 'ls #0 #1 #2'"
echo "ls a;b a b"
echo ""
echo "$G_OPT_POS_SPEC_S, $G_OPT_POS_SPEC_L <fmt-str>"
echo "Change the positional argument string. Default is '#%d', i.e. a '#'"
echo "followed by a number."
echo ""
echo "$G_OPT_SYNTAX_STR_S, $G_OPT_SYNTAX_STR_L <expected-syntax>"
echo "Syntax clarification string. E.g."
echo "$ echo a b c | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2'"
echo "prep.sh: error: file "-", line 1: \"a b c\": 2 fields expected, but got 3 instead"
echo "vs."
echo "$ echo a b c | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2' $G_OPT_SYNTAX_STR_S '<host> <port>'"
echo "prep.sh: error: file "-", line 1: \"a b c\": 2 fields expected, but got 3 instead; syntax should be \"<host> <port>\""
echo ""
echo "$G_OPT_STRING_S, $G_OPT_STRING_L <string-with-pos-args>"
echo "The string to operate on."
echo ""
echo "$G_OPT_SYNTAX_CHECK_S, $G_OPT_SYNTAX_CHECK_L <<fnum>~<regex>;<fnum>~<regex>...>"
echo "Matches the field with number <fnum> to its respective <regex>. If the"
echo "match fails, quit with an error. Effectively, this allows for regex"
echo "syntax checks. E.g."
echo "$ echo a b | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2' $G_OPT_SYNTAX_CHECK_S '1~^localhost$'"
echo "prep.sh: error: file \"-\", line 1: \"a b\": field 1 \"a\" should match \"^localhost$\""
echo "$ echo localhost b | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2' $G_OPT_SYNTAX_CHECK_S '1~^localhost$'"
echo "nc localhost b"
echo "$ echo localhost b | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2' $G_OPT_SYNTAX_CHECK_S '1~^localhost$;2~^[0-9]+$'"
echo "prep.sh: error: file \"-\", line 1: \"localhost b\": field 2 \"b\" should match \"^[0-9]+$\""
echo "$ echo localhost 8000 | $G_SCRIPT_NAME $G_OPT_STRING_S 'nc #1 #2' $G_OPT_SYNTAX_CHECK_S '1~^localhost$;2~^[0-9]+$'"
echo "nc localhost 8000"
echo ""
echo "$G_OPT_DRY_RUN_S, $G_OPT_DRY_RUN_L"
echo "Print the commands which constitute this script, but do not execute them."
echo ""
echo "$G_OPT_HELP_S, $G_OPT_HELP_L"
echo "Print this screen."
echo ""
echo "$G_OPT_VERSION_S, $G_OPT_VERSION_L"
echo "Print version information."
	exit_success
}

function main
{
	assert_main_arg_num "$@"
	get_args "$@"
	call_awk
}

function assert_main_arg_num
{
	if [ "$#" -lt 1 ]; then
		print_use
	fi
}

function call_awk
{
	if [ -z "$G_STRING" ]; then
		error_exit "'$G_OPT_STRING_L' not given"
	fi
	
	awk_it "$G_FIELD_SEP"\
		"$G_NUM_FIELDS"\
		"$G_POS_SPEC"\
		"$G_SYNTAX_STR"\
		"$G_SYNTAX_CHECK"\
		"$G_STRING"\
		"$G_CMD_LINE_OTHER"
}

function awk_it
{
	local L_FLD_SEP="$1"
	local L_FLD_NUM="$2"
	local L_POS_SPEC="$3"
	local L_SYNTAX_STR="$4"
	local L_SYNTAX_CHK="$5"
	local L_STRING_LINE="$6"
	local L_FILES="$7"

readonly local L_SRC='
function get_fld_syntax_err_str(fld_num,    err_str, res) {
	res = get_file_pos()
	res = sprintf("%s: %d fields expected, but got %d instead",
		res, get_fld_num(), fld_num)
	
	err_str = get_syntx_str()
	if (err_str)
		res = sprintf("%s; syntax should be \"%s\"", res, err_str)
		
	return res
}	

function get_rx_syntax_err_str(fld_num, fld_str, fld_rx,    res) {
	res = get_file_pos()
	res = sprintf("%s: field %d \"%s\" should match \"%s\"",
				res, fld_num, fld_str, fld_rx)
	return res
}

function get_file_pos() {
	return sprintf("file \"%s\", line %d: \"%s\"",
		FILENAME, FNR, $0)
}

function err_quit(msg,    err_msg) {
	err_msg = sprintf("%s: error: %s", get_prog_name(), msg)
	print err_msg > "/dev/stderr"
	exit(1)
}

function set_prog_name(pn) {_B_prog_name = pn}
function get_prog_name()   {return _B_prog_name}
function set_fld_num(fn)   {_B_fld_num = fn}
function get_fld_num()     {return _B_fld_num}
function set_syntx_str(sx) {_B_err_str = sx}
function get_syntx_str()   {return _B_err_str}
function set_pos_spec(pr)  {_B_pos_spec = pr}
function get_pos_spec()    {return _B_pos_spec}
function set_check_rx(fnum, crx) {_B_check_crx[fnum] = crx}
function get_check_rx(fnum)      {return _B_check_crx[fnum]}
function set_string_line(cl) {_B_string_line = cl}
function get_string_line()   {return _B_string_line}


function init() {
	set_prog_name(ProgName ? ProgName : ARGV[0])
	set_fld_num(FldNum ? FldNum : 2)
	set_pos_spec(PosRegx ? PosRegx : "#%d")
	set_syntx_str(SyntaxStr ? SyntaxStr : "")
	
	if (SyntaxChk)
		process_check_rx(SyntaxChk)
	
	if (TheString)
		set_string_line(TheString)
	else
		err_quit("-vTheString not given")
}

function CHK_RX_RS() {return ";"}
function CHK_RX_FS() {return "~"}
function CHK_RX() {return "^[0-9]+~.+$"}
function process_check_rx(chk_rx,    chk1, i, arr1, len1, arr2) {
	
	len1 = split(chk_rx, arr1, CHK_RX_RS())
	for (i = 1; i <= len1; ++i) {
		chk1 = arr1[i]
		
		if (!match(chk1, CHK_RX()))
			err_quit(sprintf("\"%s\" should match \"%s\"",
				chk1, CHK_RX()))
		
		split(chk1, arr2, CHK_RX_FS())
		set_check_rx(arr2[1], arr2[2])
	}
}

function check_field(field, fnum,    chk_rx) {
		chk_rx = get_check_rx(fnum)
		if (chk_rx && !match(field, chk_rx))
			err_quit(get_rx_syntax_err_str(fnum, field, chk_rx))
}

function gen_str(arr, len,    i, str, regx, fld) {
	str = get_string_line()
	for (i = len; i >= 0; --i) {
		fld = arr[i]
		
		check_field(fld, i)
		regx = sprintf(get_pos_spec(), i)
		gsub(regx, fld, str)
	}
	return str
}

function process_line(fname, linenum, str,    arr, len) {
	len = split(str, arr)
	arr[0] = str
	if (len != get_fld_num())
		err_quit(get_fld_syntax_err_str(len))
	return gen_str(arr, len)
}

BEGIN {init()}
$0 ~ /^[[:space:]]*$/ {next}
$0 ~ /^[[:space:]]*#/ {next}
{print process_line(FILENAME, FNR, $0)}
'	
	echo_eval "awk $L_FLD_SEP "\
		"-vProgName='$0' "\
		"-vFldNum='$L_FLD_NUM' "\
		"-vPosRegx='$L_POS_SPEC' "\
		"-vSyntaxStr='$L_SYNTAX_STR' "\
		"-vSyntaxChk='$L_SYNTAX_CHK' "\
		"-vTheString='$L_STRING_LINE' "\
		"'$L_SRC' "\
		"$L_FILES"
}

function echo_eval
{
	if [ ! -z "$G_DRY_RUN" ]; then
		echo "$@"
	else
		eval "$@"
	fi
}

main "$@"
